[SYCL] Add basic tests for non-uniform groups

Pennycook opened the pull request:
Tests for https://github.com/intel/llvm/pull/8202.

Pennycook commented:
@steffenlarsen - What's the new magic for running clang-format?  I tried this, but it doesn't seem to be obeying the expected formatting...

```bash
git diff -U0 --no-color HEAD^ | llvm/clang/tools/clang-format/clang-format-diff.py -p1 -i
```

steffenlarsen commented:
> @steffenlarsen - What's the new magic for running clang-format? I tried this, but it doesn't seem to be obeying the expected formatting...
> 
> ```shell
> git diff -U0 --no-color HEAD^ | llvm/clang/tools/clang-format/clang-format-diff.py -p1 -i
> ```

That is what I normally use, but sometimes different versions of clang-format may disagree on certain scenarios. You could try building a new version of clang-format and adding it to your PATH or just apply the suggested format by hand.

steffenlarsen commented on SYCL/NonUniformGroups/masked_sub_group.cpp:
```diff
@@ -0,0 +1,77 @@
+    sycl::accessor LeaderAcc{LeaderBuf, CGH, sycl::write_only};
+    sycl::accessor MaskAcc{MaskBuf, CGH, sycl::write_only};
+    const auto KernelFunc = [=](sycl::nd_item<1> item)
+        [[sycl::reqd_sub_group_size(32)]] {
```
We should probably check that the device supports this sub-group size and skip the test if it doesn't.

Pennycook replied:
Fixed in https://github.com/intel/llvm-test-suite/pull/1574/commits/33c11f91aa3544197315640bf9180eff21085d94.

Pennycook commented:
> You could try building a new version of clang-format and adding it to your PATH or just apply the suggested format by hand.

Good idea. Building a new version of clang-format fixed it.

steffenlarsen reviewed (approved):
LGTM!

Pennycook commented:
Converting this to draft to prevent it from being merged, same as https://github.com/intel/llvm/pull/8202.

Pennycook committed:
[SYCL] Add basic tests for non-uniform groups

Tests the ability to create an instance of each new group type,
and the correctness of the core member functions.

Signed-off-by: John Pennycook <john.pennycook@intel.com>

Pennycook commented:
@steffenlarsen, @JackAKirk - These should be fixed, now, and aligned with the latest code over at https://github.com/intel/llvm/pull/8202.

Pennycook committed:
Fix typo in -fsyntax-only flag

Signed-off-by: John Pennycook <john.pennycook@intel.com>

Pennycook committed:
Add CPU to unsupported platforms

Pennycook commented:
@cperkinsintel, @steffenlarsen: I tried to make it so these won't run on CPU, but the failure logs in https://github.com/intel/llvm/pull/8202 suggest it's still failing on the CPU.  Have I got the UNSUPPORTED syntax wrong?

steffenlarsen commented:
> @cperkinsintel, @steffenlarsen: I tried to make it so these won't run on CPU, but the failure logs in [intel/llvm#8202](https://github.com/intel/llvm/pull/8202) suggest it's still failing on the CPU. Have I got the UNSUPPORTED syntax wrong?

Syntax looks right and since you're only using `GPU_RUN_PLACEHOLDER` it should not even try to run these on CPU. It may have just been a CI bug so I've restarted the verification on your PR.

Pennycook committed:
Disable optimizations in tangle_group test

cperkinsintel commented:
These tests are for a new feature. They have successfully run on the feature PR here: https://github.com/intel/llvm/pull/8202
